python/ray/dag/compiled_dag_node.py

```diff
@@ -510,139 +524,262 @@ def reset_and_wait_intermediate_future(self) -> Any:
+            input_data = self.reset_and_wait_intermediate_future()
+            try:
+                _process_return_vals(input_data, return_single_output=False)
+                # Wait for any future in the input data.
+                for i in range(len(input_data)):
+                    val = input_data[i]
+                    if isinstance(val, DAGOperationFuture):
+                        resolved_future = val.wait()
+                        # The only source of future is NCCL recv.
+                        # The future wraps around a one-element list.
+                        assert isinstance(resolved_future, list)
+                        assert len(resolved_future) == 1
+                        resolved_future = resolved_future[0]
+                        # The resolved value can be an exception.
+                        if isinstance(resolved_future, RayTaskError):
+                            raise resolved_future.as_instanceof_cause()
+                        input_data[i] = resolved_future
+
+                resolved_inputs = []
+                for task_input in self.task_inputs:
+                    resolved_inputs.append(task_input.resolve(input_data))
+
+                if self.sync_group is not None:
+                    # Run a synchronous NCCL operation.
+                    method = self.sync_group.execute
+                else:
+                    # Run an actor method.
+                    method = getattr(class_handle, self.method_name)
+                try:
+                    output_val = method(*resolved_inputs, **self.resolved_kwargs)
+                except Exception as exc:
+                    output_val = _wrap_exception(exc)
+
+                # When overlap_gpu_communication is enabled, wrap the result in a
+                # GPUFuture so that this compute operation can be overlapped
+                # with communication.
+                self.wrap_and_set_intermediate_future(
+                    output_val, wrap_in_gpu_future=overlap_gpu_communication
+                )
+            except Exception as exc:
+                # Previous task raised an application-level exception.
+                # Propagate it and skip the actual task. We don't need to wrap the
+                # exception in a RayTaskError here because it has already been wrapped
+                # by the previous task.
+                self.wrap_and_set_intermediate_future(
+                    exc, wrap_in_gpu_future=overlap_gpu_communication
+                )
+
+        with self._send_stream:
+            # To overlap GPU communication for NCCL recv, write the future as output to
+            # the downstream task, which waits on the future in its compute operation.
+            if self.requires_nccl_read and overlap_gpu_communication:
+                output_val = self._intermediate_future
+                self._intermediate_future = None
+            else:
+                output_val = self.reset_and_wait_intermediate_future()
```

Comment:
Comment: Write of base `ClassMethodNode` and `NcclSendNode` needs to wait on the compute future.